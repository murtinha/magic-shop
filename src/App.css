.wrapper {
  background-color: #F8F8F8;
  position: fixed;
  width: 40vw;
  height: 40vh;
  top: 8%;
  right: 6%;
  z-index: 2;
  transition: width 0.1s ease-in-out;
  border: 1px solid #F8F8F8;
  box-shadow: 4px solid black;
  display: flex;
  flex-wrap: wrap;
  flex-direction: column;
}

.appWrapper {
  display: flex;
  padding: 16px;
  flex-wrap: wrap;
  justify-content: center;
}

.deck {
  padding: 4px;
  height: 20px;
  cursor: pointer;
}

.deck:hover {
  text-decoration: underline;
}

.appHeader {
  padding: 8px;
  border-bottom: 1px solid black;
  width: 100%;
  margin-bottom: 8px;
}

.icon {
  cursor: pointer;
  float: right;
  font-size: 25px;
}


