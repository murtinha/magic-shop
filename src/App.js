import React, { Component } from 'react';
import { Portal } from 'react-portal';
import './App.css';
import Card from './components/card';

class App extends Component {
  constructor() {
    super();
    this.state = {
      cards: [],
      deck: [],
      isOpen: false,
    }
  }

  componentDidMount() {
    fetch('https://api.magicthegathering.io/v1/cards?page=28')
      .then(res => res.json())
      .then(data => this.setState({ cards: data.cards }))
  }
    
  addCard = card => {
    const { deck } = this.state;
    this.setState({ deck: [...deck, card] });
  }

  removeCard = id => {
    const { deck } = this.state;
    const idx = deck.findIndex(card => card.cardId === id);
    deck.splice(idx, 1);
    this.setState({ deck });
  }

  toggleOpen= () => {
   this.setState({ isOpen: !this.state.isOpen });
} 

  render() {
    const { deck, isOpen } = this.state;
    return (
      <div className="appWrapper">
        <div className="appHeader">
          MONTE SEU DECK DE COMMANDER!
          <span className="icon" onClick={this.toggleOpen}>
            Meu deck: {deck.length}
          </span>
        </div>
        { 
          this.state.cards.map((card, idx) => <Card key={idx} card={card} onAdd={this.addCard} onRemove={this.removeCard} />) 
        }
        {
          isOpen &&
          <Portal>
            <div className="wrapper">
              {
                deck.map((c,idx) => <div className="deck" key={idx}> - {c.name} </div>)
              }
              </div>
          </Portal>
        }
      </div>
    );
  }
}

export default App;
