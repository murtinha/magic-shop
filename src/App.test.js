import React from 'react';
import { shallow } from 'enzyme';
import App from './App';

describe('<App />', () => {
  const cards = [
    {
      image: 'path/imagem.png',
      name: 'Black Lotus',
      id: 1,
      manaCost: '0',
      type: 'Artifact',
      rarity: 'Rare',
      setName: 'Magic collection',
      text: 'Card text',
      flavor: 'Card flavor',
      artist: 'murtinha',
    },
    {
      image: 'path/imagem2.png',
      name: 'Bufao',
      id: 2,
      manaCost: 'B2G',
      type: 'Creature',
      rarity: 'Legend',
      setName: 'Magic collection 2',
      text: 'Card text 2',
      flavor: 'Card flavor 2',
      artist: 'murtinha2',
    }
  ];
  const deck = [
    {
      name: 'Black Lotus',
      cardId: 1,
    },
    {
      name: 'Bufao',
      cardId: 2,
    }
  ];
  const renderShallowWrapper = () => shallow(<App />);

  describe('render', () => {
    it('should render correctly', () => {
      const wrapper = renderShallowWrapper();
      expect(wrapper).toMatchSnapshot();
    });

    it('should render correctly with cards', () => {
      const wrapper = renderShallowWrapper();
      wrapper.setState({ cards });
      expect(wrapper).toMatchSnapshot();
    });

    it('should render correctly opened', () => {
      const wrapper = renderShallowWrapper();
      wrapper.setState({ isOpen: true, deck });
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    const mockFetch = global.fetch;

    afterEach(() => {
      global.fetch = mockFetch;
    });

    it('should fetch cards on did mount', () => {
      global.fetch = jest.fn().mockImplementation(() => Promise.resolve());
      const wrapper = renderShallowWrapper();
      const url = 'https://api.magicthegathering.io/v1/cards?page=28';
      wrapper.instance().componentDidMount();

      expect(fetch).toHaveBeenCalledWith(url);
    });

    it('should add card properly', () => {
      const wrapper = renderShallowWrapper();
      const card = { name: 'Forest', cardId: 3 };
      wrapper.setState({ deck }); 
      wrapper.instance().addCard(card);
      expect(wrapper.state('deck')).toEqual([
        ...deck,
        card,
      ]);
    });

    it('should remove card properly', () => {
      const wrapper = renderShallowWrapper();
      wrapper.setState({ deck }); 
      wrapper.instance().removeCard(2);
      expect(wrapper.state('deck')).toEqual([
        deck[0],
      ]);
    });

    it('should toggle state properly', () => {
      const wrapper = renderShallowWrapper();
      wrapper.instance().toggleOpen();
      expect(wrapper.state('isOpen')).toBeTruthy();
      wrapper.instance().toggleOpen();
      expect(wrapper.state('isOpen')).toBeFalsy();
    });
  });
});
