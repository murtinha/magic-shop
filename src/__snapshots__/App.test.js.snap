// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> render should render correctly 1`] = `
<div
  className="appWrapper"
>
  <div
    className="appHeader"
  >
    MONTE SEU DECK DE COMMANDER!
    <span
      className="icon"
      onClick={[Function]}
    >
      Meu deck: 
      0
    </span>
  </div>
</div>
`;

exports[`<App /> render should render correctly opened 1`] = `
<div
  className="appWrapper"
>
  <div
    className="appHeader"
  >
    MONTE SEU DECK DE COMMANDER!
    <span
      className="icon"
      onClick={[Function]}
    >
      Meu deck: 
      2
    </span>
  </div>
  <Portal>
    <div
      className="wrapper"
    >
      <div
        className="deck"
        key="0"
      >
         - 
        Black Lotus
         
      </div>
      <div
        className="deck"
        key="1"
      >
         - 
        Bufao
         
      </div>
    </div>
  </Portal>
</div>
`;

exports[`<App /> render should render correctly with cards 1`] = `
<div
  className="appWrapper"
>
  <div
    className="appHeader"
  >
    MONTE SEU DECK DE COMMANDER!
    <span
      className="icon"
      onClick={[Function]}
    >
      Meu deck: 
      0
    </span>
  </div>
  <Card
    card={
      Object {
        "artist": "murtinha",
        "flavor": "Card flavor",
        "id": 1,
        "image": "path/imagem.png",
        "manaCost": "0",
        "name": "Black Lotus",
        "rarity": "Rare",
        "setName": "Magic collection",
        "text": "Card text",
        "type": "Artifact",
      }
    }
    key="0"
    onAdd={[Function]}
    onRemove={[Function]}
  />
  <Card
    card={
      Object {
        "artist": "murtinha2",
        "flavor": "Card flavor 2",
        "id": 2,
        "image": "path/imagem2.png",
        "manaCost": "B2G",
        "name": "Bufao",
        "rarity": "Legend",
        "setName": "Magic collection 2",
        "text": "Card text 2",
        "type": "Creature",
      }
    }
    key="1"
    onAdd={[Function]}
    onRemove={[Function]}
  />
</div>
`;
